import { findOneUser, updateUserById, findSingleUserByUsername, findUsersByUsername} from "../services/userService";
import { NextFunction, Response } from "express";
import { omit } from "lodash";
import { customRequest } from "../types/customDefinition";
import { ApiError } from "../util/ApiError";
const omitData = ["password"];

export const updateUser = async (
  req: customRequest,
  res: Response,
  next: NextFunction
) => {
  try {
    const { id: userId } = req.user;

    let body = req.body;
    body = omit(body, omitData);

    const user = await findOneUser({ id: userId });

    if (!user) {
      throw new ApiError(400, "User not found");
    }

    const updated = await updateUserById(body, parseInt(userId, 10));

    return res.status(200).json({
      updated: updated[0],
      msg: updated[0] ? "Data updated successfully" : "failed to update",
      error: false,
    });
  } catch (err) {
    next(err);
  }
};

export const findAllUsersByUsername = async (
  req: customRequest,
  res: Response,
  next: NextFunction
) => {
  try {
    const { id: userId } = req.user;

    let body = req.body;
    body = omit(body, omitData);
    if (!body.username) {
      throw new ApiError(400, "Please provide username");
    }

    const user = await findOneUser({ id: userId });

    if (!user) {
      throw new ApiError(400, "User not found");
    }

    const users = await findUsersByUsername({ username: body.username });

    return res.status(200).json({
      updated: users,
      msg: "users fetched successfully",
      error: false,
    });
  } catch (err) {
    next(err);
  }
};

export const getUserData = async (
  req: customRequest,
  res: Response,
  next: NextFunction
) => {
  try {
    return res.status(200).json({
      data: req.user,
      error: false,
    });
  } catch (err) {
    next(err);
  }
};
